### ISSUE TITLE
pin numbering of pin headers is wrong

### ISSUE THREAD

**Issue by davidperrenoud** [fritzing/fritzing-app#2199]
_From [th.gauwe...@googlemail.com](https://code.google.com/u/114319454448349078647/) on September 06, 2012 15:02:58_

The pin numbering of the pin headers do not follow the convention.

Actually its anti-counterclockwise.
The convention is one row to the next row.

E.g. pin 1 is in the corner upper left.
Then pin 2 is below pin 1.
But the convention is that pin 2 is in the same row in the right from pin 1.

see http://www.kreatives-chaos.com/artikel/avr-programmieradapter for example.

_Original issue: http://code.google.com/p/fritzing/issues/detail?id=2199_

**Comment by Zalewa** [comment]
Hi. I've just stumbled upon this problem. Are there any plans on fixing this? If you don't want to break compatibility with existing schematics then perhaps it should be provided as an option that would be enabled by default for new schematics? I was actually looking for such option before I found out it's actually a bug.

**Comment by CCommander** [comment]
Another year has passed and it occurs that I have stumbled upon this problem. Are you still aware of the issue? Is there a perspective that this could be improved?

**Comment by Kahldar** [comment]
Hi ! Is there a plan to fix this bug ? Pins numbers on shrouded headers are still wrong (0.9.3). Thanks.

**Comment by davidgiven** [comment]
Hello --- any news on this one?

**Comment by th-gauweiler** [comment]
Hopefully the new team can solve the issue.

**Comment by EsGeh** [comment]
Yeah, this is an annoying problem!
Maybe the backwards compatibility issue could be solved by just by providing a fixed version as a separate part so existing sketches don't break...?
We are all looking forward to this being fixed soon!

**Comment by KjellMorgenstern** [comment]
Hi there, thanks for pinging this issue and bringing it to my attention.  Adding a new version of a part is not that complicated, so the fact this is still unresolved is a hint that there is a bit more to this issue. 

If someone could add screenshots or even an .svg of how the part / feature should look like this would already be of big help to me.

**Comment by th-gauweiler** [comment]
The look does not have to change.

The point is: the concrete part will be generated on the fly. Just the
algorithm how the pins are numbered is wrong. I guess it is the same as for
IC.

Kjell <notifications@github.com> schrieb am Sa., 11. Mai 2019, 19:39:

**Comment by th-gauweiler** [comment]
Assume you have a 2x3 connector and three parts.
The pinning of the connector should be in this form:
one     two
three   four
five      six
Now connect two with the resistor.
four with the capacitor and
six with the diode.
schematic.fzz shows the right schematic.
And this should lead to the connection in file pcb.fzz.

The example was made with a now fresh installed frizzing version.
Version 0.9.3b <http://fritzing.org/download/0.9.3b/> was released on Juni
2, 2016.

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virenfrei.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Am Sa., 11. Mai 2019 um 23:42 Uhr schrieb Thomas Gauweiler <
th.gauweiler@googlemail.com>:

**Comment by th-gauweiler** [comment]
>If someone could add screenshots

I just see, that my example files, which are appended to my email are
filtered out by GitHub.

How may I append screenshot or even files to an issue?

**Comment by KjellMorgenstern** [comment]
Drag and drop usually works on GitHub

th-gauweiler <notifications@github.com> schrieb am Mi. 15. Mai 2019 um
12:55:

**Comment by th-gauweiler** [comment]
[wrong pinning.zip](https://github.com/fritzing/fritzing-app/files/3213105/wrong.pinning.zip)

An example, that illustrate the example:

Assume you have a 2x3 connector and three parts.
The pinning of the connector should be in this form:
one     two
three   four
five      six
Now connect two with the resistor.
four with the capacitor and
six with the diode.
schematic.fzz shows the right schematic.
And this should lead to the connection in file pcb.fzz.

Its an error in generating a concrete item of a generic part.
The parameter double row must be set.

**Comment by KjellMorgenstern** [comment]
From schematic.fzz
![image](https://user-images.githubusercontent.com/6494025/58273970-199e1b00-7d92-11e9-8863-187db5bfcb9f.png)
![image](https://user-images.githubusercontent.com/6494025/58274028-3df9f780-7d92-11e9-8613-965462bb1cf3.png)

From pcb.fzz
![image](https://user-images.githubusercontent.com/6494025/58274123-7b5e8500-7d92-11e9-9e79-d95dbc89e6a9.png)
![image](https://user-images.githubusercontent.com/6494025/58274151-8a453780-7d92-11e9-9615-801dc94dabb2.png)

**Comment by KjellMorgenstern** [comment]
So, currently you get
```
1 6
2 5
3 4
```

Looking at an Arduino Mega (they have the odd numbers on the right), one would expect:
```
 ...
 22 21
 24 23
 26 25
 ...
```
or following the numbering on the raspberry pi connector
```
1 2
3 4
5 6
...
```
I understand, that the Rasperry PI connector naming scheme is closest to what is suggested in this thread.

Besides (another issue, which I might have closed a bit to early), it can be a bit confusing that you have a single line in the Breadboard view and a double line in the PCB.

**Comment by failiz** [comment]
Revive imported label

**Comment by failiz** [comment]
I do not think that changing the algorithm is very difficult. The problem is how to keep backward compatibility (do not modify previous sketches which were done with the current algorithm). 
@KjellMorgenstern , are the generated parts stored in the sketch file or are they generated again when you open the sketch? If the latter, is there a way to check if the connectors were done with some specific version? 
If the connectors are not generated again, we could replace the code without problem. If not, we should check the version of the connector and use one algorithm or the other. 
Are there any other approaches?

**Comment by KjellMorgenstern** [comment]
There could be a new property
'pin order' ->
'left-right', 'right-left', 'counter clockwise',
 'clockwise', 'down', 'up'

For simplicity, I don't think I have only seen lr, rl, and ccw , so those three might be sufficient.
Default could then be ccw, so old parts have the same behavior, no matter if they are stored or generated.
If generated, the parts would not be backward compatible, but is see that currently files like 
'generic_female_alternating_pin_header_6_100mil.fzp' are created, so it should be possible
to add a this as
generic_shrouded_pin_header_10_100mil.fzp ->
generic_shrouded_pin_header_10_100mil_ccw.fzp

and then
generic_shrouded_pin_header_10_100mil_rl.fzp
generic_shrouded_pin_header_10_100mil_lr.fzp
...

**Comment by KjellMorgenstern** [comment]
Since that is quite a number of options, and not all combinations are possible,  I'd consider adding a button in the inspector that shows a dialog, possibly with a preview.

**Comment by failiz** [comment]
It looks that there is no standard:
https://electronics.stackexchange.com/questions/106426/standard-nomenclature-for-2-row-connector-numbering-schemes
https://www.cirris.com/learning-center/general-testing/special-topics/27-connector-pin-counting-sequence

However, I agree that the zig-zag pattern is more common (specially for ribbon cables).
I am ok with making it an option, but the default one should be the 'left-right' (zig-zag). We could use the suffix of the file as you suggested to determine the option used. And if there is no suffix, use the ccw one. 
I would also limit the options to just ccw  and 'left-right'. I not sure if it makes sense to have more (the benefit compared to the time to implement them).

**Comment by th-gauweiler** [comment]
10 years

**Comment by louigi600** [comment]
2023 and still not fix and you get charged money for he latest version of Fritzing while they cant even get right a generic 16pin 2x8 for ribbon cable pin header numbering. This is so disappointing.
but there are correct ones for 2x5 and 2x10... but unlike the gereric one for these you can't set an arbitrary pin count.

I really need 2x8 with sholder male with classical ribbon cable numbering: anyone made one ?

**Comment by john-clark** [comment]
2 more years. It would be really nice if the part editor would allow things to be changed more readily.
